### PR TITLE
Enable querying constants and value kinds

### PR DESCRIPTION
**Pull request by tbennun**
This PR adds bindings for obtaining whether a value is constant, what its `ValueKind` is, and its exact constant value in certain cases (integer, floating point; in other cases the IR string is returned instead).

This is useful for inspecting constant values in Python code.

### CODE REVIEW

**Comment by tbennun** [comment]
This PR also fixes #727

**Review by apmasell — COMMENTED** [review]
Thanks! This is a useful feature. Please add unit tests.

File: ffi/value.cpp [hunk 1]
```diff
@@ -309,6 +309,34 @@ LLVMPY_DisposeOperandsIter(LLVMOperandsIteratorRef GI) {
     delete llvm::unwrap(GI);
 }
 
+API_EXPORT(LLVMBool)
```
[the remark this answers is not in the document]
Comment (tbennun) [reply ffi/value.cpp]: Fixed, thank you!

**Comment by tbennun** [comment]
@apmasell Removed LLVMBool from Python-facing API, added tests, and fixed formatting.

**Review by apmasell — APPROVED** [review]
Looks good. Thank you!

File: llvmlite/binding/value.py [hunk 2]
```diff
@@ -140,6 +177,14 @@ def is_instruction(self):
     def is_operand(self):
         return self._kind == 'operand'
 
+    @property
+    def is_constant(self):
```
Comment (sklam) [inline llvmlite/binding/value.py:181]: Needs doc update in docs/source/user-guide/binding/type-references.rst

File: llvmlite/binding/value.py [hunk 3]
```diff
@@ -140,6 +177,14 @@ def is_instruction(self):
...
+        return bool(ffi.lib.LLVMPY_IsConstant(self))
+
+    @property
+    def value_kind(self):
```
Comment (sklam) [inline llvmlite/binding/value.py:185]: Needs doc update in docs/source/user-guide/binding/type-references.rst

File: llvmlite/binding/value.py [hunk 4]
```diff
@@ -299,6 +344,43 @@ def opcode(self):
                              % (self._kind,))
         return ffi.ret_string(ffi.lib.LLVMPY_GetOpcodeName(self))
 
+    def get_constant_value(self):
```
Comment (sklam) [inline llvmlite/binding/value.py:347]: Needs doc update in docs/source/user-guide/binding/type-references.rst

File: llvmlite/binding/value.py [hunk 5]
```diff
@@ -299,6 +344,43 @@ def opcode(self):
...
+            if accuracy_loss.value:
+                warnings.warn(
+                    'Accuracy loss encountered in conversion of constant '
+                    f'value {str(self)}'
+                )
+
+            return value
+
+        # Otherwise, return the IR string
+        return str(self)
```
Comment (sklam) [inline llvmlite/binding/value.py:382]: Needs to add a test for the str return

File: llvmlite/binding/value.py [hunk 6]
```diff
@@ -299,6 +344,43 @@ def opcode(self):
...
+                ('little' if little_endian.value else 'big'),
+                signed=False,
+            )
+        elif self.value_kind == ValueKind.constant_fp:
+            # Convert floating-point values to double (Python float)
+            accuracy_loss = c_bool(False)
+            value = ffi.lib.LLVMPY_GetConstantFPValue(self,
+                                                      byref(accuracy_loss))
+            if accuracy_loss.value:
+                warnings.warn(
```
Comment (sklam) [inline llvmlite/binding/value.py:374]: Needs to add test for the warning

File: llvmlite/binding/value.py [hunk 7]
```diff
@@ -299,6 +344,43 @@ def opcode(self):
                              % (self._kind,))
         return ffi.ret_string(ffi.lib.LLVMPY_GetOpcodeName(self))
 
+    def get_constant_value(self):
+        """
+        Return the constant value, either as a literal (when supported)
+        or as a string.
+        """
+        if not self.is_constant:
+            raise ValueError('expected constant value, got %s'
```
Comment (sklam) [inline llvmlite/binding/value.py:353]: Needs to add a test for the raise

**Review by sklam — CHANGES_REQUESTED** [review]
I have some additional concerns mainly the lack of docs for the new code, and the lack of test coverage for some cases.

File: llvmlite/binding/value.py (hunk 6 above)
Comment (tbennun) [reply llvmlite/binding/value.py:374]: @sklam For some reason, despite how much I try, I can't seem to trigger the `losesInfo` boolean with a single constant (maybe this appears during passes). Not sure how to proceed here, any suggestions?

**Comment by tbennun** [comment]
@sklam Added documentation and more tests (apart from testing the warning, which I couldn't reproduce but is part of the LLVM interface, see above comment). I also noticed that adding one argument to `get_constant_value` significantly improves its usage for signed integers, which was apparent in one of the tests.

**Comment by tbennun** [comment]
Any updates on this @sklam ? This is not waiting for me as of two months ago.

**Comment by esc** [comment]
@tbennun sorry, seems like we have dropped the ball on this one. I am changing the label now and will ping @sklam again. It is now scheduled for inclusion in the 0.41 release to be tagged during the week starting 7th of August.

File: llvmlite/binding/value.py (hunk 6 above)
Comment (sklam) [reply llvmlite/binding/value.py:379]: Sorry for the delay. 

After some digging into the LLVM source, the floating-point constants are encoded as double-precision so there will not be any information loss here. To force info loss, we will need to use bigger types like the long-double. For example: `%const = fadd fp128 0xLF3CB1CCF26FBC178452FB4EC7F91DEAD, 0xLF3CB1CCF26FBC178452FB4EC7F91973F`. This will require converting the 128-bit float to 64-bit float thus triggering the warning. 

However, this brings to my attention that a warning is not appropriate here. This API is not giving user a choice to read float bigger than double-precision.  In the LLVM OCAML binding, the similar function will return `None` (https://github.com/llvm/llvm-project/blob/1defa781243f9d0bc66719465e4de33e9fb7a243/llvm/bindings/ocaml/llvm/llvm_ocaml.c#L1030-L1032). I think the Python API should raise an exception.

Also, maybe it is better to have separate functions for int constant and float constant, so that the float version can provide more control on what to do if information is lost. For instance, if user is okay with the info loss, they can force the downcast.

**Review by sklam — CHANGES_REQUESTED** [review]
Thanks for adding to the docs. Those looks great.

However, I have followed the rabbit hole and discovered more details to be considered for `get_constant_value` and the info loss issue. See below.

File: llvmlite/binding/value.py (hunk 6 above)
Comment (tbennun) [reply llvmlite/binding/value.py:379]: That's great, I agree and already modified the constant float test. Only thing I'm not sure about is the separate functions for int and float constants. One of the main benefits of Python is that you can return multiple types. Maybe this should be another keyword argument, and keyword args should act like preferences?

I am already extending the behavior to support other types of constants (e.g. structs), and it was beneficial to be able to call the same method on the nested data: https://github.com/tbennun/llvmlite/commit/3992ea175c164c9deedb8644b12b138c2e0c0dfc#diff-27ab06b9fa73e854c37e5ccf261269588cd8e02fadbd71c788102bc46c4a2f1bR390

Comment (tbennun) [reply llvmlite/binding/value.py]: @sklam all changes are complete and the PR is ready for re(re)review.

**Review by sklam — APPROVED** [review]
